### PR TITLE
golangci: Enable errorlint

### PR DESCRIPTION
**Pull request by jrajahalme**
Enable errorlint in golangci-lint to catch uses of improper formatters for Go errors. This helps avoid unnecessary error/warning logs that cause CI flakes, when benign error cases are not caught due to failing error unwrapping when a string or value formatter has been used instead of the dedicated `%w`.

Note to reviewers: Only `fmt.Errorf()` calls should have been changed to use `%w`. Other changes to use `errors.Is()` and `errors.As()` to replace comparison and type cast, respectively should be reviewed carefully.

Note to backporters: In addition to resolving conflicts, it is possible that the release branches have more code that needs to be revised. Use `make lint` locally to find any additional code sites that need to be fixed. Unfortunately the `fix` options seems to not be implemented for `errorlint`.

Related: #31147

### CODE REVIEW

**Review by giorio94 — COMMENTED** [review]
One comment inline, looks good otherwise for my codeowners. I'll approve once out of draft, as it otherwise doesn't count now.

File: daemon/cmd/cni/config.go
Comment (giorio94) [inline daemon/cmd/cni/config.go]: I don't know why the comment got lost. However, the changes in this file should be reverted, as the error is being logged. Even better would be using the dedicated `WithError` method provided by logrus, but that's possibly out of scope for this PR.

Comment (jrajahalme) [reply daemon/cmd/cni/config.go]: noticed, reverted :-)

**Review by giorio94 — APPROVED** [review]
The changes for my codeowners look good. Thanks

**Review by michi-covalent — APPROVED** [review]

File: bugtool/cmd/configuration.go
Comment (christarazi) [inline bugtool/cmd/configuration.go]:
```suggestion
		return fmt.Errorf("Cannot marshal config: %w", err)
```

File: bugtool/cmd/configuration.go
Comment (christarazi) [inline bugtool/cmd/configuration.go]:
```suggestion
		return fmt.Errorf("Cannot write config: %w", err)
```

File: clustermesh-apiserver/etcdinit/root.go
Comment (christarazi) [inline clustermesh-apiserver/etcdinit/root.go:175]:
```suggestion
			if errors.As(err, &exec.ExitError{}) {
```

File: daemon/cmd/endpoint.go
Comment (christarazi) [inline daemon/cmd/endpoint.go:141]:
```suggestion
		if errors.As(err, &api.APIError{}) {
```

Comment (christarazi) [reply daemon/cmd/endpoint.go:141]: There's a few more of these. Feel free to inline all of these nits. Not sure why the linter would generate code like this (assuming that it did).

File: pkg/alibabacloud/eni/node.go
Comment (christarazi) [inline pkg/alibabacloud/eni/node.go]:
```suggestion
			fmt.Errorf("%s: %w", errUnableToGetSecurityGroups, err)
```

There's a few more in this file that need the same change (I won't call out each one), otherwise it becomes difficult to tell which part of the error msg is from the error or if it's the context around it.

File: pkg/aws/eni/node.go
Comment (christarazi) [inline pkg/aws/eni/node.go]:
```suggestion
			fmt.Errorf("%s: %w", errUnableToGetSecurityGroups, err)
```

More in this file as well.

**Review by christarazi — COMMENTED** [review]
Thanks for the PR! Good to have this enabled.

From the lines that I've called out, there are much more instances that I did not call out for the sake repetitiveness, but I think it's worth going through and applying the suggestions to all instances while we're at it.

File: operator/pkg/gateway-api/grpcroute_reconcile.go
Comment (sayboras) [inline operator/pkg/gateway-api/grpcroute_reconcile.go:142]: nice, I didn't know that we can wrap multiple errors at all.

**Review by sayboras — APPROVED** [review]
Looks good for files in operator/gateway-api

File: bugtool/cmd/configuration.go
Comment (tommyp1ckles) [inline bugtool/cmd/configuration.go:237]: Although these error %w format changes are usually pretty innocent, I am a little bit concerned about the number of such changes at once.   It is possible for error formatting changes to affect logic elsewhere, such as when matching on context.Cancelled.

File: daemon/cmd/endpoint.go
Comment (jrajahalme) [reply daemon/cmd/endpoint.go:141]: It did not, but in this case the second parameter really needs to be a pointer to pointer, as it needs to be a pointer to  a type that implements the `error` interface, and in this case it is the pointer to `api.APIError` that does that, not `api.APIError` itself. IMO it is clearer to separate the two in different lines. Besides, the variable needs to be named if it is used in the following code (here it is not).

Comment (jrajahalme) [reply daemon/cmd/endpoint.go:141]: See https://github.com/golang/go/issues/37625

File: clustermesh-apiserver/etcdinit/root.go
Comment (jrajahalme) [reply clustermesh-apiserver/etcdinit/root.go:175]: See response below.

File: bugtool/cmd/configuration.go
Comment (jrajahalme) [reply bugtool/cmd/configuration.go]: Done :-)

File: bugtool/cmd/configuration.go
Comment (jrajahalme) [reply bugtool/cmd/configuration.go]: Done.

File: pkg/alibabacloud/eni/node.go
Comment (jrajahalme) [reply pkg/alibabacloud/eni/node.go]: Done.

File: pkg/aws/eni/node.go
Comment (jrajahalme) [reply pkg/aws/eni/node.go]: Done.

File: bugtool/cmd/configuration.go
Comment (jrajahalme) [reply bugtool/cmd/configuration.go:237]: This change makes broken matching using `errors.Is()` on `context.Cancelled` work (see #31147). Are you aware of matching that would be broken (rather than fixed) by this change?

**Comment by jrajahalme** [comment]
Rebased due to #31398 having merged. #31398 did the part that showed up in #31147.

File: operator/pkg/gateway-api/grpcroute_reconcile.go
Comment (jrajahalme) [reply operator/pkg/gateway-api/grpcroute_reconcile.go:142]: Works since Go 1.20 :-)

**Review by christarazi — APPROVED** [review]

**Review by bimmlerd — APPROVED** [review]

**Review by dylandreimerink — APPROVED** [review]

**Review by harsimran-pabla — APPROVED** [review]

**Review by tklauser — APPROVED** [review]

**Review by nathanjsweet — APPROVED** [review]

**Review by ysksuzuki — APPROVED** [review]

**Review by mhofstetter — APPROVED** [review]

**Review by glibsm — APPROVED** [review]

**Review by squeed — APPROVED** [review]

File: pkg/k8s/slim/k8s/apis/labels/selector.go
Comment (squeed) [inline pkg/k8s/slim/k8s/apis/labels/selector.go:654]: This should probably be dropped; "slim" is supposed to be a vendored copy with some code removed.

**Review by squeed — CHANGES_REQUESTED** [review]
One question about slim selector.

**Review by tommyp1ckles — APPROVED** [review]

File: pkg/k8s/slim/k8s/apis/labels/selector.go
Comment (jrajahalme) [reply pkg/k8s/slim/k8s/apis/labels/selector.go:654]: meaning that we should alter the generated error here, or anywhere under `pkg/k8s/slim`? We'll have to add a comment/annotation to skip linting either inline here, or globally for the `pkg/k8s/slim` in the Makefile.

Comment (aanm) [reply pkg/k8s/slim/k8s/apis/labels/selector.go:654]: @squeed it's ok, the "slim" code is supposed to be a vendored copy yes but we do have already some modifications targeted for Cilium. It's not a 1:1 copy anymore.

**Review by squeed — APPROVED** [review]

File: pkg/k8s/slim/k8s/apis/labels/selector.go
Comment (squeed) [reply pkg/k8s/slim/k8s/apis/labels/selector.go:654]: ack!

**Review by julianwiedmann — APPROVED** [review]

**Pull request merged**
